aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -0,0 +1,507 @@
+    if input.follow_symlinks() {
+        walker = walker.follow_links(true);
+    }
+    if !input.recursive() {
+        walker = walker.max_depth(1);
+    }
+    walker
+}
+
+fn derive_object_key<'a>(
+    relative_filename: &'a str,
+    object_key_prefix: Option<&str>,
+    object_key_delimiter: Option<&str>,
+) -> Result<Cow<'a, str>, error::Error> {
+    if let Some(delim) = object_key_delimiter {
+        if delim != DEFAULT_DELIMITER && relative_filename.contains(delim) {
+            return Err(error::invalid_input(format!(
+                "a custom delimiter `{delim}` should not appear in `{relative_filename}`"
+            )));
+        }
+    }
+
+    let delim = object_key_delimiter.unwrap_or(DEFAULT_DELIMITER);
+
+    let relative_filename = if delim.starts_with(MAIN_SEPARATOR) {
+        Cow::Borrowed(relative_filename)
+    } else {
+        Cow::Owned(relative_filename.replace(MAIN_SEPARATOR, delim))
+    };
+
+    let object_key = if let Some(prefix) = object_key_prefix {
+        if prefix.ends_with(delim) {
+            Cow::Owned(format!("{prefix}{relative_filename}"))
+        } else {
+            Cow::Owned(format!("{prefix}{delim}{relative_filename}"))
+        }
+    } else {
+        relative_filename
+    };
+
+    Ok(object_key)
+}
+
+pub(super) async fn upload_objects(
+    ctx: UploadObjectsContext,
+    work_rx: Receiver<Result<UploadObjectJob, error::Error>>,
+) -> Result<(), error::Error> {
+    while let Ok(job) = work_rx.recv().await {
+        match job {
+            Ok(job) => {
+                let key = job.key.clone();
+                let result = upload_single_obj(&ctx, job).await;
+                match result {
+                    Ok(bytes_transferred) => {
+                        ctx.state.successful_uploads.fetch_add(1, Ordering::SeqCst);
+
+                        ctx.state
+                            .total_bytes_transferred
+                            .fetch_add(bytes_transferred, Ordering::SeqCst);
+
+                        tracing::debug!("worker finished uploading object {:?}", key);
+                    }
+                    Err(err) => {
+                        tracing::debug!("worker failed to upload object {:?}: {}", key, err);
+                        handle_failed_upload(err, &ctx, Some(key))?;
+                    }
+                }
+            }
+            Err(err) => handle_failed_upload(err, &ctx, None)?,
+        }
+    }
+
+    tracing::trace!("req channel closed, worker finished");
+    Ok(())
+}
+
+async fn upload_single_obj(
+    ctx: &UploadObjectsContext,
+    job: UploadObjectJob,
+) -> Result<u64, error::Error> {
+    let UploadObjectJob { object, key } = job;
+
+    // `object` gets consumed by `input` so calculate the content length in advance
+    let bytes_transferred: u64 = object
+        .size_hint()
+        .upper()
+        .ok_or_else(crate::io::error::Error::upper_bound_size_hint_required)
+        .unwrap();
```

Comment:
**Important:** The upper size_hint might not be equal to the actual bytes transferred.

Reply:
Comment added in 0c4f872